internal/service/news/news.go

```diff
@@ -0,0 +1,144 @@
+package news
+
+import (
+	gocontext "context"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/newUser1337/task-news/internal/config"
+	"github.com/newUser1337/task-news/internal/context"
+	"github.com/newUser1337/task-news/internal/entity"
+	"github.com/newUser1337/task-news/internal/repository"
+)
+
+const articlesListEndpoint string = "https://www.htafc.com/api/incrowd/getnewlistinformation?count=50"
+const articleEndpoint string = "https://www.htafc.com/api/incrowd/getnewsarticleinformation?id=%d"
+
+type articleRepository interface {
+	SaveArticleList(gocontext.Context, []*entity.NewsletterNewsItemDB) ([]int64, error)
+	SaveArticle(gocontext.Context, *entity.NewsArticleDB) (int64, error)
+}
+
+type NewsServer struct {
+	appCtx  *context.AppCtx
+	article articleRepository
+}
+
+func StartNewNewsServer(appCtx *context.AppCtx, cfg *config.Config, rep *repository.Repository) {
+	newsServer := &NewsServer{
+		appCtx:  appCtx,
+		article: rep.Article,
+	}
+
+	go newsServer.maintain(cfg)
+}
+
+func (n *NewsServer) maintain(cfg *config.Config) {
+	n.appCtx.WgAdd(1)
+	defer n.appCtx.WgDone()
+
+	refreshTime := cfg.Core.RefreshInterval
+	timer := time.NewTimer(refreshTime)
+
+	if err := n.getArticels(); err != nil {
+		err = fmt.Errorf("news-srv: failed to get articels %s", err)
+		log.Println(err)
+	}
+
+	for {
+		timer.Reset(refreshTime)
+		select {
+		case <-n.appCtx.GetGoContext().Done():
+			if !timer.Stop() {
+				<-timer.C
+			}
+			return
+		case <-timer.C:
+		}
+
+		if err := n.getArticels(); err != nil {
+			err = fmt.Errorf("news-srv: failed to get articels %s", err)
+			log.Println(err)
+		}
+	}
+}
+
+func (n *NewsServer) getArticels() error {
+	articlesExt := new(entity.ArticleListExternal)
```

Comment:
`new` usually indicates there is some non-standard behaviour here to pay attention to - https://dave.cheney.net/2014/08/17/go-has-both-make-and-new-functions-what-gives

```
As a code reviewer, the use of new, like the use of named return arguments, is a signal that the code is trying to do something clever and I need to pay special attention. It may be that code really is clever, but more than likely, it can be rewritten to be clearer and more idiomatic.
```

better have `var articlesExt entity.ArticleListExternal` and use `&articlesExt` in `fetchData`

Reply:
fixed

Reply:
fixed